### PR TITLE
Allow the nameservers to be populated from systemd-resolve

### LINKED ISSUE #57618: [BUG] Grain does not report configured nameservers for systemd-resolved
**Description**
On systems using `systemd-resolved`, by default the `/etc/resolv.conf` file lists 127.0.0.53 as the nameserver (for clients not using the libc or bus APIs). The list of actual nameservers is available e.g. via `systemd-resolve --status`, or `resolvectl dns` on newer systems.

I'm not sure when `resolvectl` was added. It's not available in `237` but is in `245`.

**Setup**
A default install of e.g. Ubuntu 18.04 (server).

**Steps to Reproduce the behavior**
```
# salt '*' grains.get dns
minion:
    ----------
    domain:
    ip4_nameservers:
        - 127.0.0.53
    ip6_nameservers:
    nameservers:
        - 127.0.0.53
    options:
        - edns0
    search:
        - local
    sortlist:
```
```
# systemd-resolve --status | tail

Link 2 (ens3)
      Current Scopes: DNS
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: yes
         DNS Servers: 1.1.1.1
                      1.0.0.1
          DNS Domain: local
```

**Expected behavior**
The system's actual configured nameservers should be returned.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3000.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
         Jinja2: 2.10
        libgit2: 0.26.0
       M2Crypto: 0.32.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.26.2
         Python: 3.6.9 (default, Apr 18 2020, 01:56:04)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 16.0.2
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5

System Versions:
           dist: Ubuntu 18.04 bionic
         locale: ISO-8859-1
        machine: x86_64
        release: 4.15.0-101-generic
         system: Linux
        version: Ubuntu 18.04 bionic
```
</details>

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Allow the nameservers to be populated from systemd-resolve.

### What issues does this PR fix or reference?
Fixes: #57618

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/utils/dns.py [hunk 1]
```diff
@@ -1216,5 +1204,67 @@ def parse_resolv(src="/etc/resolv.conf"):
...
+
+    nameservers = []
+    ip4_nameservers = []
+    ip6_nameservers = []
+    domain = ""
+
+    cmd = "systemd-resolve --status"
+    cmd = __salt__["cmd.run_all"](
+        "{}".format(cmd), python_shell=False, output_loglevel="quiet"
+    )
```
Comment (s0undt3ch) [inline salt/utils/dns.py:1224]: Why not:
```
cmd = ["systemd-resolve", "--status"]
    cmd = __salt__["cmd.run_all"](
        "{}".format(cmd), python_shell=True, output_loglevel="quiet"
    )
```

File: salt/utils/dns.py [hunk 2]
```diff
@@ -1216,5 +1204,67 @@ def parse_resolv(src="/etc/resolv.conf"):
...
+
+    if cmd["retcode"] == 0:
+        prev_section = None
+        lines = [re.sub("[\t ]", "", line) for line in cmd["stdout"].splitlines()]
```
Comment (s0undt3ch) [inline salt/utils/dns.py:1228]: I do not know the output for an example, but my guess is that `systemd-resolve` doesn't support a parseable output, nor a plain `.strip()` would do it?
Wouldn't `.replace("\t", "").replace(" ", "")` achieve the same? Or is `re.sub` just faster?

File: salt/utils/dns.py [hunk 3]
```diff
@@ -1216,5 +1204,67 @@ def parse_resolv(src="/etc/resolv.conf"):
...
+
+        contents = {}
+        prev_section = None
+        for line in lines:
+            section = None
+            if ":" in line and line.count(":") <= 1:
```
Comment (s0undt3ch) [inline salt/utils/dns.py:1234]: If `:` is in line, then there will always be one, so, `and line.count(":") == 1`?

File: salt/utils/dns.py [hunk 4]
```diff
@@ -1216,5 +1204,67 @@ def parse_resolv(src="/etc/resolv.conf"):
...
+                        version = ip_addr.version
+                        ip_addr = str(ip_addr)
+                        if ip_addr not in nameservers:
+                            nameservers.append(ip_addr)
+                        if version == 4 and ip_addr not in ip4_nameservers:
+                            ip4_nameservers.append(ip_addr)
+                        elif version == 6 and ip_addr not in ip6_nameservers:
+                            ip6_nameservers.append(ip_addr)
+                    except ValueError as exc:
+                        log.error("%s: %s", addr, exc)
```
Comment (s0undt3ch) [inline salt/utils/dns.py:1262]: This usually hides valuable information on the traceback. Let's at least add `exc_info_on_loglevel=logging.DEBUG`

**Comment by s0undt3ch** [comment]
https://stackoverflow.com/questions/52589322/generating-yaml-by-parsing-systemd-resolve-output#comment92123238_52589322
❓

File: salt/utils/dns.py (hunk 1 above)
Comment (garethgreenaway) [reply salt/utils/dns.py:1224]: i thought we wanted to avoid using `python_shell=True`?

File: salt/utils/dns.py (hunk 2 above)
Comment (garethgreenaway) [reply salt/utils/dns.py:1228]: Unfortunately it doesn't have an option to output into a parseable format like JSON.  I'll test with using `.replace` and see how that works.

File: salt/utils/dns.py (hunk 3 above)
Comment (garethgreenaway) [reply salt/utils/dns.py:1234]: Yup.  Already made that change, just need to commit and push.

File: salt/utils/dns.py (hunk 4 above)
Comment (garethgreenaway) [reply salt/utils/dns.py:1262]: 👍

**Comment by garethgreenaway** [comment]
@s0undt3ch Totally missed that comment.  That will make this a lot easier.  Thanks!

File: salt/utils/dns.py (hunk 1 above)
Comment (s0undt3ch) [reply salt/utils/dns.py:1224]: Hmm, we do. Doesn't help that we use `python_shell` and subprocess uses `shell`.

Also, what I really wanted to write was
```python
cmd = ["systemd-resolve", "--status"]
cmd = __salt__["cmd.run_all"](cmd, python_shell=False, output_loglevel="quiet")
```

Comment (s0undt3ch) [reply salt/utils/dns.py:1224]: Ie, pass the command as a list, not as a string.

Comment (garethgreenaway) [reply salt/utils/dns.py]: Since we can use a different, auto generated resolv.conf file don't need this code.

**Review by s0undt3ch — APPROVED** [review]

**Pull request merged**